fix: expose provider visibility and kms endpoint drift

piyush117 opened the pull request:
### Description

fixes: #911

Exposes `provider_visibility` across all pattern entry points and introduces a dedicated `kms_endpoint_type` variable to control the endpoint used for KMS resource API calls (key rings, keys, key policies).

**Problem**

When `provider_visibility` is set to `private`, KMS resource API calls were still using the `public` endpoint (hardcoded default), causing Schematics workspace tests to fail with forced replacements on `ibm_kms_key_rings` / `ibm_kms_key_policies` and in-place updates on `ibm_kms_key`.

**Fix**

A dedicated `kms_endpoint_type` variable is exposed with a default of `private`. This mirrors the approach used in [terraform-ibm-kms-all-inclusive](https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/blob/12cc695af20a867fc630f3cad88703a3b5f3dd9c/solutions/fully-configurable/main.tf#L18) and gives users explicit control independent of provider visibility.

Changes:

- `patterns/dynamic_values/config_modules/key_management` — generated keys now set `endpoint = null`, allowing the module-level `kms_endpoint_type` default to apply via `coalesce`
- `kms/main.tf` — all 3 KMS resources use `coalesce(each.value.endpoint, var.key_management.kms_endpoint_type)` so per-key endpoint overrides still work
- `kms/variables.tf`, `variables.tf` (root) — `kms_endpoint_type` field added to `key_management` object type (default `private`)
- `patterns/vsi/module`, `patterns/roks/module`, `patterns/vpc/module` — accept `kms_endpoint_type` and pass it into `key_management`; removed the now-unused `provider_visibility` variable from these internal modules
- `patterns/vsi`, `patterns/roks`, `patterns/vpc`, `patterns/mixed`, `patterns/vsi-quickstart`, `patterns/roks-quickstart` — declare `kms_endpoint_type` (default `private`) and thread it through
- `ibm_catalog.json` — added `kms_endpoint_type` (hidden) to all 5 working directory flavor entries
- Tests — set `kms_endpoint_type = "public"` alongside `provider_visibility = "public"` in all PR and HPCS test setups

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

Exposes `provider_visibility` to control IBM provider network visibility across all Landing Zone pattern entry points. Adds a dedicated `kms_endpoint_type` variable (default `private`) to explicitly set the endpoint type used for KMS API calls, decoupled from provider visibility. Users running in private network environments no longer need manual workarounds to prevent KMS resource drift.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [x] If relevant, a test for the change is included or updated with this PR.
- [x] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

piyush117 committed:
Expose provider visibility

piyush117 commented:
/run pipeline

piyush117 committed:
SKIP UPGRADE TEST

piyush117 committed:
SKIP UPGRADE TEST

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
SKIP UPGRADE TEST

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

Ak-sky committed:
Merge branch 'main' into 11323-expose-provider-visibility

Ak-sky commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

Ak-sky committed:
Merge branch 'main' into 11323-expose-provider-visibility

Ak-sky committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 committed:
Merge branch 'main' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 committed:
Merge branch '11323-expose-provider-visibility' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 committed:
small fix

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

jor2 reviewed (commented):
Straightforward addition of `provider_visibility` across all patterns. Looks good overall, one note on the catalog defaults.

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented on patterns/mixed/variables.tf:
```diff
@@ -8,6 +8,16 @@ variable "ibmcloud_api_key" {
+  description = "Set the visibility value for the IBM terraform provider. Supported values are `public`, `private`, `public-and-private`. [Learn more](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/guides/custom-service-endpoints)."
+  type        = string
+  default     = "private"
+  validation {
```
Should that be the case for all the patterns where variable `provider_visibility` is declared?

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 committed:
merge conflicts

piyush117 committed:
merge conflicts

piyush117 committed:
merge conflicts

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 committed:
introduced a new kms_endpoint_type field

piyush117 committed:
Merge branch 'main' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 committed:
Merge branch '11323-expose-provider-visibility' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 committed:
minor fix

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 committed:
UNSKIP UPGRADE TEST

piyush117 committed:
Merge branch 'main' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 committed:
Merge branch '11323-expose-provider-visibility' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 committed:
Merge branch 'main' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 committed:
Merge branch '11323-expose-provider-visibility' of github.com:terraform-ibm-modules/terraform-ibm-landing-zone into 11323-expose-provider-visibility

piyush117 commented:
tested the upgrade in schematics,
<img width="1728" height="1117" alt="Screenshot 2026-05-12 at 12 52 36 AM" src="https://github.com/user-attachments/assets/a574b61d-5d7e-441a-9352-1f8528b78a7e" />
<img width="1728" height="1117" alt="Screenshot 2026-05-12 at 12 52 23 AM" src="https://github.com/user-attachments/assets/3996a1bd-1bbd-4f84-bd87-0dd40bbe96ec" />

piyush117 committed:
trivy and go-git fix

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
merge

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

Aashiq-J reviewed (commented):
Instead of using the provider visibility to get the kms_endpoint type, use the same approach as we use in [here](https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/blob/12cc695af20a867fc630f3cad88703a3b5f3dd9c/solutions/fully-configurable/main.tf#L18)

Set the default to private and if user wants they can change it.

piyush117 committed:
feat: decouple kms_endpoint_type from provider_visibility

Add a dedicated `kms_endpoint_type` variable (default `private`) to all
pattern entry points instead of deriving the KMS endpoint type from
`provider_visibility`. Mirrors the approach used in terraform-ibm-kms-all-inclusive.

- Remove unused `provider_visibility` from module/ subdirs (tflint fix)
- Add `kms_endpoint_type` to ibm_catalog.json for all working directories

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
fix: add kms_endpoint_type to override-example and set public in test

The override-example routes through vsi/module which now defaults
kms_endpoint_type to private. CI runners don't have private network
access, so the test must explicitly pass public.

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

imprateeksh committed:
Merge branch 'main' into 11323-expose-provider-visibility

imprateeksh commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 commented:
/run pipeline

piyush117 committed:
Merge branch 'main' into 11323-expose-provider-visibility

piyush117 commented:
/run pipeline

Aashiq-J reviewed (commented):
why are all the tests using public endpoint?

Aashiq-J commented:
/run pipeline

Aashiq-J reviewed (approved):
LGTM

piyush117 commented:
/run pipeline

piyush117 commented:
> why are all the tests using public endpoint?

- Schematics tests call `setupOptionsSchematics` → set their own `TerraformVars` as a `[]testschematic.TestSchematicTerraformVar` list — completely separate from the regular `setupOptions*` functions.

- Neither `provider_visibility` nor `kms_endpoint_type` appear in any Schematics `TerraformVars` list, so both default to `"private"`. Schematics runs in IBM Cloud's private network and can reach the private KMS endpoint. 

- Regular terratests call `setupOptionsVsiPattern`, `setupOptionsRoksPattern`, etc., which explicitly set both to `"public"`. These tests skip when `enableSchematicsTests = true`.

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 8.21.6 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone/releases/tag/v8.21.6)
- `v8.21.6`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
